src/Catalog/Helpers/GalleryDatabaseHelper.cs

```diff
@@ -0,0 +1,116 @@
+                "Created");
+        }
+
+        public Task<SortedList<DateTime, IList<FeedPackageDetails>>> GetPackagesEditedSince(DateTime since, int top)
+        {
+            return GetPackagesInOrder(
+                since,
+                top,
+                package => package.LastEditedDate,
+                "LastEdited");
+        }
+
+        /// <summary>
+        /// Returns a <see cref="SortedList{DateTime, IList{FeedPackageDetails}}"/> from the gallery database.
+        /// </summary>
+        /// <param name="keyDateFunc">The <see cref="DateTime"/> field to sort the <see cref="FeedPackageDetails"/> on.</param>
+        private async Task<SortedList<DateTime, IList<FeedPackageDetails>>> GetPackagesInOrder(
+            DateTime since,
+            int top,
+            Func<FeedPackageDetails, DateTime> keyDateFunc,
+            string columnName)
+        {
+            var allPackages = await GetPackages(since, top, columnName);
+
+            return OrderPackagesByKeyDate(allPackages, keyDateFunc);
+        }
+
+
+        /// <summary>
+        /// Returns a <see cref="SortedList{DateTime, IList{FeedPackageDetails}}"/> of packages.
+        /// </summary>
+        /// <param name="keyDateFunc">The <see cref="DateTime"/> field to sort the <see cref="FeedPackageDetails"/> on.</param>
+        public static SortedList<DateTime, IList<FeedPackageDetails>> OrderPackagesByKeyDate(
+            IReadOnlyCollection<FeedPackageDetails> packages,
+            Func<FeedPackageDetails, DateTime> keyDateFunc)
+        {
+            var result = new SortedList<DateTime, IList<FeedPackageDetails>>();
+
+            foreach (var package in packages)
+            {
+                var packageKeyDate = keyDateFunc(package);
+                if (!result.TryGetValue(packageKeyDate, out IList<FeedPackageDetails> packagesWithSameKeyDate))
+                {
+                    packagesWithSameKeyDate = new List<FeedPackageDetails>();
+                    result.Add(packageKeyDate, packagesWithSameKeyDate);
+                }
+
+                packagesWithSameKeyDate.Add(package);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Asynchronously gets a <see cref="IList{FeedPackageDetails}"/> from the gallery database.
+        /// </summary>
+        /// <param name="connectionFactory">Provides a connection to the gallery database.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{TResult}.Result" />) returns an
+        /// <see cref="IList{FeedPackageDetails}" />.</returns>
+        private async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackages(
+            DateTime since,
+            int top,
+            string columnName)
+        {
+            using (var sqlConnection = await _connectionFactory.OpenAsync())
+            using (var sqlTransaction = sqlConnection.BeginTransaction())
```

Comment:
We are doing two DB queries now (previous one via feed). How we we know this transaction is giving us the proper level of isolation and therefore self consistency? We need a load test? I am not an expert in isolation levels or SQL lock escalation so I can't see what I need to see to be convinced that this is working okay.

Reply:
Currently, the default READ COMMITTED isolation level is applied. I think, however, that we should be using SNAPSHOT isolation: https://docs.microsoft.com/en-us/sql/t-sql/statements/set-transaction-isolation-level-transact-sql?view=sql-server-2017

I'm no expert either which is why it is good to get this reviewed ;)
Maybe @dtivel wants to have a say in this?

Reply:
No more need for read-only transaction; now is a single SQL query.